charger/peblar.go

```diff
@@ -0,0 +1,304 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// Details on the Peblar modbus server obtained from: https://developer.peblar.com/modbus-api
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/modbus"
+	"github.com/volkszaehler/mbmd/encoding"
+)
+
+// Peblar charger implementation
+type Peblar struct {
+	log     *util.Logger
+	conn    *modbus.Connection
+	curr    uint32
+	enabled bool
+	phases  uint16
+}
+
+const (
+	// Meter addresses
+	peblarEnergyTotalAddress   = 30000
+	peblarSessionEnergyAddress = 30004
+	peblarPowerPhase1Address   = 30008
+	peblarPowerPhase2Address   = 30010
+	peblarPowerPhase3Address   = 30012
+	peblarPowerTotalAddress    = 30014
+	peblarVoltagePhase1Address = 30016
+	peblarVoltagePhase2Address = 30018
+	peblarVoltagePhase3Address = 30020
+	peblarCurrentPhase1Address = 30022
+	peblarCurrentPhase2Address = 30024
+	peblarCurrentPhase3Address = 30026
+
+	// Config addresses
+	peblarSerialNumberAddress  = 30050
+	peblarProductNumberAddress = 30062
+	peblarFwIdentifierAddress  = 30074
+	peblarPhaseCountAddress    = 30092
+	peblarIndepRelayAddress    = 30093
+
+	// Control addresses
+	peblarCurrentLimitSourceAddress = 30112
+	peblarCurrentLimitActualAddress = 30113
+	peblarModbusCurrentLimitAddress = 40000
+	peblarForce1PhaseAddress        = 40002
+
+	// Diagnostic addresses
+	peblarCpStateAddress = 30110
+)
+
+func init() {
+	registry.Add("peblar", NewPeblarFromConfig)
+}
+
+//go:generate go run ../cmd/tools/decorate.go -f decoratePeblar -b *Peblar -r api.Charger -t "api.PhaseSwitcher,Phases1p3p,func(int) error"
+
+// NewPeblarFromConfig creates a Peblar charger from generic config
+func NewPeblarFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := modbus.TcpSettings{
+		ID: 255,
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewPeblar(cc.URI, cc.ID)
+}
+
+// NewPeblar creates Peblar charger
+func NewPeblar(uri string, id uint8) (api.Charger, error) {
+	conn, err := modbus.NewConnection(uri, "", "", 0, modbus.Tcp, id)
+	if err != nil {
+		return nil, err
+	}
+
+	log := util.NewLogger("peblar")
+	conn.Logger(log.TRACE)
+
+	// Register contains the physically connected phases
+	b, err := conn.ReadInputRegisters(peblarPhaseCountAddress, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	wb := &Peblar{
+		log:    log,
+		conn:   conn,
+		curr:   6000,                       // assume min current
+		phases: binary.BigEndian.Uint16(b), // TODO why do we need this?
```

Comment:
why do we need this?

Reply:
This is read during init to know how much voltage/current registers must be read by the getPhaseValues function.

If a phase is not connected, the modbus read will return an error, as the value not available (garbage). Therefore read the configured/physically connected amount once here and not each time getPhaseValues is called.